drivers/soundwire/amd_init.c

```diff
@@ -0,0 +1,168 @@
+					   unsigned int num_res,
+					   const void *data,
+					   size_t size_data)
+{
+	pdevinfo->name = name;
+	pdevinfo->id = id;
+	pdevinfo->parent = parent;
+	pdevinfo->num_res = num_res;
+	pdevinfo->res = res;
+	pdevinfo->data = data;
+	pdevinfo->size_data = size_data;
+	pdevinfo->fwnode = fw_node;
+}
+
+static struct sdw_amd_ctx *sdw_amd_probe_controller(struct sdw_amd_res *res)
+{
+	struct sdw_amd_ctx *ctx;
+	struct acpi_device *adev;
+	struct resource *sdw_res;
+	struct acp_sdw_pdata *sdw_pdata;
+	struct platform_device_info pdevinfo[2];
+	u32 link_mask;
+	int count, index;
+
+	if (!res)
+		return NULL;
+
+	adev = acpi_fetch_acpi_dev(res->handle);
+	if (!adev)
+		return NULL;
+
+	if (!res->count)
+		return NULL;
+
+	count = res->count;
+	dev_dbg(&adev->dev, "Creating %d SDW Link devices\n", count);
+
+	/*
+	 * we need to alloc/free memory manually and can't use devm:
+	 * this routine may be called from a workqueue, and not from
+	 * the parent .probe.
+	 * If devm_ was used, the memory might never be freed on errors.
+	 */
+	ctx = kzalloc(sizeof(*ctx), GFP_KERNEL);
+	if (!ctx)
+		return NULL;
+
+	ctx->count = count;
+	ctx->link_mask = res->link_mask;
+	sdw_res = kzalloc(sizeof(*sdw_res), GFP_KERNEL);
+	if (!sdw_res) {
+		kfree(ctx);
+		return NULL;
+	}
+	sdw_res->flags = IORESOURCE_MEM;
+	sdw_res->start = res->addr;
+	sdw_res->end = res->addr + res->reg_range;
+	memset(&pdevinfo, 0, sizeof(pdevinfo));
+	sdw_pdata = kzalloc(sizeof(*sdw_pdata) * 2, GFP_KERNEL);
+	if (!sdw_pdata) {
+		kfree(sdw_res);
+		kfree(ctx);
+		return NULL;
+	}
+	link_mask = ctx->link_mask;
+	for (index = 0; index < count; index++) {
+		if (!(link_mask & BIT(index)))
+			continue;
+
+		sdw_pdata[index].instance = index;
+		sdw_pdata[index].acp_sdw_lock = res->acp_lock;
+		acp_sdw_fill_platform_dev_info(&pdevinfo[index], res->parent,
+					       acpi_fwnode_handle(adev), "amd_sdw_manager",
+					       index, sdw_res, 1, &sdw_pdata[index],
+					       sizeof(struct acp_sdw_pdata));
+		ctx->pdev[index] = platform_device_register_full(&pdevinfo[index]);
+		if (IS_ERR(ctx->pdev[index]))
+			goto err;
+	}
+	kfree(sdw_pdata);
+	kfree(sdw_res);
+	return ctx;
+
+err:
+	while (index--) {
+		if (!(link_mask & BIT(index)))
+			continue;
+
+		platform_device_unregister(ctx->pdev[index]);
+	}
+
+	kfree(sdw_pdata);
+	kfree(sdw_res);
+	kfree(ctx);
+	return NULL;
+}
+
+struct sdw_amd_ctx *sdw_amd_probe(struct sdw_amd_res *res)
+{
+	return sdw_amd_probe_controller(res);
+}
+EXPORT_SYMBOL_NS(sdw_amd_probe, SOUNDWIRE_AMD_INIT);
+
+int sdw_amd_startup(struct sdw_amd_ctx *ctx)
+{
+	struct amd_sdw_manager *amd_manager;
+	int i, ret;
+
+	/* Startup SDW Manager devices */
+	for (i = 0; i < ctx->count; i++) {
+		if (!(ctx->link_mask & BIT(i)))
+			continue;
+		amd_manager = dev_get_drvdata(&ctx->pdev[i]->dev);
+		ret = amd_sdw_manager_start(amd_manager);
+		if (ret)
+			return ret;
+	}
+
+	return 0;
+}
+EXPORT_SYMBOL_NS(sdw_amd_startup, SOUNDWIRE_AMD_INIT);
```

Comment:
You need to explain why this is needed.

In the case of Intel, we had to start the link after the firmware download on older platforms, so we had a three step process

a) ACPI parsing and device creation without any interaction with hardware
b) hardware probe
c) SoundWire startup

Are b) and c) really different steps in your case, i.e. do you need the startup at all?

Reply:
yes startup is required. provided justification for startup in one of the replies for your review comment

Reply:
In our case also hardware probe and SoundWire startup is different.